daemon/access.go

```diff
@@ -0,0 +1,143 @@
+	"strconv"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/polkit"
+)
+
+type accessResult int
+
+const (
+	accessOK accessResult = iota
+	accessUnauthorized
+	accessForbidden
+	accessCancelled
+)
+
+var polkitCheckAuthorization = polkit.CheckAuthorization
+
+func checkPolkitAction(r *http.Request, ucred *ucrednet, action string) accessResult {
+	var flags polkit.CheckFlags
+	allowHeader := r.Header.Get(client.AllowInteractionHeader)
+	if allowHeader != "" {
+		if allow, err := strconv.ParseBool(allowHeader); err != nil {
+			logger.Noticef("error parsing %s header: %s", client.AllowInteractionHeader, err)
+		} else if allow {
+			flags |= polkit.CheckAllowInteraction
+		}
+	}
+	// Pass both pid and uid from the peer ucred to avoid pid race
+	if authorized, err := polkitCheckAuthorization(ucred.pid, ucred.uid, action, nil, flags); err == nil {
+		if authorized {
+			// polkit says user is authorised
+			return accessOK
+		}
+	} else if err == polkit.ErrDismissed {
+		return accessCancelled
+	} else {
+		logger.Noticef("polkit error: %s", err)
+	}
+	return accessUnauthorized
+}
+
+// accessChecker checks whether a particular request is allowed.
+//
+// An access checker will either allow a request, deny it, or return
+// accessUnknown, which indicates the decision should be delegated to
+// the next access checker.
+type accessChecker interface {
+	checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult
+}
+
+type openAccess struct{}
+
+func (ac openAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	// Forbid access from snapd-snap.socket
+	if ucred != nil && ucred.socket == dirs.SnapSocket {
+		return accessForbidden
+	}
+
+	return accessOK
+}
+
+// authenticatedAccess allows requests from authenticated users.
+//
+// A user is considered authenticated if they provide a macaroon, are
+// the root user according to peer credentials, or granted access by
+// Polkit.
+type authenticatedAccess struct {
+	Polkit string
+}
+
+func (ac authenticatedAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	// Forbid access from snapd-snap.socket
+	if ucred != nil && ucred.socket == dirs.SnapSocket {
+		return accessForbidden
+	}
+
+	if user != nil {
+		return accessOK
+	}
+
+	// Next checks require ucred
+	if ucred == nil {
+		return accessUnauthorized
+	}
+
+	if ucred.uid == 0 {
+		return accessOK
+	}
+
+	if ac.Polkit != "" {
+		return checkPolkitAction(r, ucred, ac.Polkit)
+	}
+
+	return accessUnauthorized
+}
+
+// rootAccess allows requests from the root uid, provided they
+// were not received on snapd-snap.socket
+type rootAccess struct{}
+
+func (ac rootAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	if ucred != nil && ucred.uid == 0 && ucred.socket != dirs.SnapSocket {
```

Comment:
Again, same request. If `ucred == nil` we should fail immediately. Having done that we can check for the socket type and then finally `ucred.uid == 0`. This will effectively state the pre-conditions very strongly and not combine them into one logical expression.